src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceCodeRefactoringProvider.RenameNamespaceCodeAction.cs

```diff
@@ -0,0 +1,295 @@
+                var targetNamespace = _state.TargetNamespace;
+
+                var declarationRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var container = declarationRoot.DescendantNodes().FirstOrDefault(node => node is TNamespaceDeclarationSyntax) ?? declarationRoot;
+
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var declaredSymbols = _service.GetDeclaredSymbolsInContainer(semanticModel, container, cancellationToken);
+
+                var documentSet = ImmutableHashSet.Create(document);
+                var editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+
+                var refLocationsInCurrentDocument = SpecializedCollections.EmptyEnumerable<ReferenceLocation>();
+                var refLocationsInOtherFilePath = SpecializedCollections.EmptyEnumerable<ReferenceLocation>();
+
+                var solution = document.Project.Solution;
+
+                // TODO: 
+                // Need to figure out how to handle MTFM projects and linked files properly.
+                // 
+                // For the document triggered the refactoring, we try to change the namespace 
+                // declaration *only in current context*, the "find reference" is invoked only 
+                // on symbols from current context as well.
+                //
+                // And for documents referencing those symbols, they can be linked/MTFM documents themselves.
+                // In which case, we treat them as independent documents and fix them individually, and at the 
+                // end rely on existing conflict resolving mechanism to merge those fixes in the same file together.
+
+                foreach (var declaredSymbol in declaredSymbols)
+                {
+                    var refSymbols = await SymbolFinder.FindReferencesAsync(
+                      declaredSymbol,
+                      solution,
+                      cancellationToken).ConfigureAwait(false);
+
+                    var refLocationsForSymbol = refSymbols.Where(refSymbol => refSymbol.Definition == declaredSymbol).SelectMany(refSymbol => refSymbol.Locations);
+
+                    // Ignore other documents with identical file path as triggering document (i.e. other TFM in a MTFM project)
+                    refLocationsInCurrentDocument = refLocationsInCurrentDocument.Concat(refLocationsForSymbol.Where(loc => loc.Document.Id == document.Id));
+                    refLocationsInOtherFilePath = refLocationsInOtherFilePath.Concat(refLocationsForSymbol.Where(loc => !PathUtilities.PathsEqual(loc.Document.FilePath, document.FilePath)));
+                }   
+
+                // Transform current document(s) first
+                solution = (await FixDeclarationDocumentAsync(document, refLocationsInCurrentDocument, cancellationToken).ConfigureAwait(false)).Project.Solution;
```

Comment:
please don't use await in a neted fashion like this.  it's too complex.  (and it wraps too long).